sdk/src/main/scala/com/horizen/SidechainState.scala

```diff
@@ -296,6 +336,25 @@ class SidechainState private[horizen] (stateStorage: SidechainStateStorage,
       (params.maxWBsAllowed * (getWithdrawalEpochInfo.lastEpochIndex + numberOfMainchainBlockReferenceInBlock)) / (params.withdrawalEpochLength - 1))
   }
 
+  private def checkCrosschainMessagesBoxesAllowed(mod: SidechainBlock, boxInThisBlock: Int): Unit = {
+    val alreadyMined = getAlreadyMinedCrosschainMessagesInCurrentEpoch
+    val mainchainBlockReferenceInBlock = mod.mainchainBlockReferencesData.size
+    val allowed = getAllowedCrosschainMessageBoxes(mainchainBlockReferenceInBlock, CryptoLibProvider.sc2scCircuitFunctions.getMaxMessagesPerCertificate)
+    val total = alreadyMined + boxInThisBlock
+    if (total > allowed) {
+      throw new IllegalStateException(s"Exceeded the maximum number of CrosschainMessages allowed!")
+    }
+  }
+
+  def getAlreadyMinedCrosschainMessagesInCurrentEpoch: Int = {
+    stateStorage.getCrossChainMessagesPerEpoch(getWithdrawalEpochInfo.epoch).size
+  }
+
```

Comment:
In the UTXO model we can have more CrossChainMessage boxes in the mempool then the ones allowed per epoch. We need to add a rule inside the forger to prevent the creation of blocks that contain more boxes then the max limit per epoch or we can run into the creation of invalid blocks.